src/Build/Evaluation/ConditionEvaluator.cs

```diff
@@ -124,9 +126,38 @@ internal static void UpdateConditionedPropertiesTable
             }
         }
 
-        // An array of hashtables with cached expression trees for all the combinations of condition strings 
-        // and parser options
-        private static volatile Hashtable[] s_cachedExpressionTrees = new Hashtable[(int)ParserOptions.AllowAll + 1];
+        // Implements a pool of expression trees for each condition.
+        // This is because an expression tree is a mutually exclusive resource (has non thread safe state while it evaluates)
+        // During high demand, new expression trees are added to the pool.
```

Comment:
More explanation of the mechanism for the adding (if all are currently busy, create a new one and push it onto the stack).